components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/WebExtensionToolbarFeature.kt

```diff
@@ -0,0 +1,80 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.toolbar
+
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import mozilla.components.browser.state.selector.findCustomTabOrSelectedTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.webextension.BrowserAction
+import mozilla.components.concept.toolbar.Toolbar
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifAnyChanged
+
+typealias WebExtensionBrowserAction = BrowserAction
+
+/**
+ * Web extension toolbar implementation that updates the toolbar whenever the state of web
+ * extensions changes.
+ */
+class WebExtensionToolbarFeature(
+    private val toolbar: Toolbar,
+    private var store: BrowserStore,
+    private val customTabId: String? = null
+) : LifecycleAwareFeature {
+    // This maps web extension id to [WebExtensionToolbarAction]
+    @VisibleForTesting
+    internal val webExtensionBrowserActions = HashMap<String, WebExtensionToolbarAction>()
+
+    private var scope: CoroutineScope? = null
+
+    /**
+     * Starts observing for the state of web extensions changes
+     */
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.ifAnyChanged { arrayOf(it.extensions) }
```

Comment:
Here we need to observer both the global web extension state as well as tab-specific state. So, this would need to be the same as before:
```
scope = store.flowScoped { flow ->
  flow.ifAnyChanged { arrayOf(it.selectedTab, it.extensions) }
    .collect { 
      state -> renderWebExtensionActions(state) 
     }
  }
```

The way it is written now we'd only collect changes to the global state, as we wouldn't observe tab-specific overrides.

Reply:
Note that you can no alternatively also write this as:
```
scope = store.flowScoped { flow ->
  flow.ifChanged { Pair(it.selectedTab(customTabId), it.extensions) }
    .collect { 
      state -> renderWebExtensionActions(state) 
     }
  }
```